src/features.rs

```diff
@@ -238,29 +235,29 @@ impl DenoiseFeatures {
             r[i] *= (self.ex[i] / (1e-8 + self.ep[i])).sqrt();
         }
         crate::interp_band_gain(&mut rf[..], &r[..]);
-        for i in 0..FREQ_SIZE {
-            self.x[i] += rf[i] * self.p[i];
-        }
+        let rf: &mut [f32] = &mut rf;
+        self.x += &self.p * &*rf;
```

Comment:
This will allocate a temporary buffer, right?

Reply:
This is just coercing the reference to an array to a reference to a slice. No copying. No allocations. No new buffers. It generally happens automatically, e.g.
```
fn main() {
    let array: [i32; 3] = [0, 1, 2];
    function_that_takes_slice(&array);
}

fn function_that_takes_slice(slice: &[i32]) {}
```

Our specific case needs explicit coercion because it's not obvious which implementation of `Mul` is being used. We're using the `Mul<[T] for DynRealDft` i.e. multiplying by a slice, but if we import a `Mul<[T; SIZE]> for DynRealDft`. i.e. multiplying by an array, we would change the operation being used here. A non-local change to the code would then affect the semantics of the program. This can be very confusing to debug. 

This is discussed in more detail [here](https://users.rust-lang.org/t/why-does-this-array-to-slice-coercion-not-work/39987) .

Reply:
Sorry, I wasn't clear: I was asking about the multiplication between the two arrays on the right hand side, which appears to allocate a temporary buffer before adding it to `self.x`.

My hidden agenda here is that I'd like to eventually publish a version for embedded devices, and part of that is keeping the allocation localized so that it can be easily conditionally compiled out.

Reply:
I've re-written the following:
```rust
self.x += &self.p * &*rf;
```
as
```rust
self.p *= &*rf;
self.x += &self.p;
```

The `MulAssign` and `AddAssign` implementations re-use the same buffer. The previous code would likely have allocated yes. Nice catch.

Reply:
Sorry to be nit-picky about this, but this would actually change the observable behavior, because `self.p` is getting modified and it's public...

Reply:
Updated with a variant that doesn't modify the `self.p` buffer :) 

A bit clunky since you need to do the first bin separately. This is because the `DynRealDft` struct needs to ensure that the first bin does not have a imaginary component.

Reply:
Oh interesting, I didn't realize that you were enforcing that in the API. I guess our requirements here are niche enough that it isn't worth having an escape hatch in easyfft itself.